ROCm 5.7.1 scilibs

AngryLoki opened the pull request.

thesamesam commented on dev-util/hip/files/hip-5.7.0-fix-unaligned-memcpy.patch:
```diff
@@ -0,0 +1,56 @@
+--- a/rocclr/device/rocm/rocvirtual.cpp
```
Please report this upstream and include a link in the patch.

The patch should also have an explanation contained within it.

AngryLoki replied:
Added to patch; link for backlink: https://github.com/ROCm-Developer-Tools/clr/issues/18

littlewu2508 commented on sci-libs/rocBLAS/files/rocBLAS-5.7.1-expand-isa-compatibility.patch:
```diff
@@ -0,0 +1,132 @@
+Allow rocBLAS to load the compatible kernels when running on
```
OK. I will test your ebuild on gfx1031.

By the way which version of hip are you using? 5.7.0?

littlewu2508 replied:
> By the way which version of hip are you using? 5.7.0?

OK, I see you prepare patches for hip, so I should use the patched hip-5.7.0

littlewu2508 commented:
@AngryLoki As I remember, there is a fatal issue with hpi-5.6.0/llvm-16 upgrade in Gentoo recorded at https://github.com/ROCmSoftwarePlatform/rocBLAS/issues/1350 and https://github.com/ROCmSoftwarePlatform/rocFFT/issues/439, so I did not bump hip to that version. I will verify if these issue is gone with 5.7.0

Also I have test enabled hip-5.6.0.ebuild at https://github.com/littlewu2508/gentoo/blob/rocm-toolchains-5.6/dev-util/hip/hip-5.6.0.ebuild (which also detects the fatal float16 issue). I'll try merging the test functionality after this PR.

AngryLoki commented:
> fatal float16 issue

@littlewu2508 Yes, you are right. I've tested with `-march=native` (efficiently, `-march=znver4` for me). With this flag float16 tests work correctly (at least some of simple selected test, which code I understand, like `test/test_log_normal_distribution`). But as you wrote about flags, I tested without `-march` and it definitely fails for all fp16 tests. Looks like software emulation for _Float16 fails (and native AVX512FP operations work). 

I've also found some places where I forgot to expand ISA for `defined(__gfx801__) || defined(__gfx1030__) || defined(__gfx1031__) || ... <and ISA-compatible offload-archs are missing>`. And then tests are working with `HSA_OVERRIDE_GFX_VERSION=10.3.0` there and fail without.

I will search for a way to defeat this `HSA_OVERRIDE_GFX_VERSION` variable to automatically load compatible ISA, when available.

littlewu2508 commented:
About `HSA_OVERRIDE_GFX_VERSION`, shall we borrow the patch from Debian?

https://salsa.debian.org/rocm-team/rocm-hipamd/-/blob/master/debian/patches/0025-improve-rocclr-isa-compatibility-check.patch

https://salsa.debian.org/rocm-team/rocm-hipamd/-/blob/master/debian/patches/0026-extend-hip-isa-compatibility-check.patch

And also proposals from the author: https://github.com/ROCmSoftwarePlatform/rocBLAS/pull/1251#issuecomment-1530822881

AngryLoki commented:
@littlewu2508 
> About `HSA_OVERRIDE_GFX_VERSION`, shall we borrow the patch from Debian that enables isa expansion? By that way user don't need to set this env var.

I added these patches and they worked like a charm for my gfx1036!
There are still rare crashes in various tests; I'll check if I can fix them. Looks like upstream developers rarely test libs with AVX512 flags and it requires some maintenance.

littlewu2508 commented:
Running `USE=test emerge =dev-libs/rocm-comgr-5.7.1` I got

```
FAILED: test/source/square.bc /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test/source/square.bc 
cd /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test && /opt/gentoo/usr/lib/llvm/17/bin/clang-17 -c --offload-arch=gfx906 -emit-llvm -fgpu-rdc --gpu-bundle-output /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr/test/source/square.hip -o source/square.bc
```

littlewu2508 commented:
> Running `USE=test emerge =dev-libs/rocm-comgr-5.7.1` I got
> 
> ```
> FAILED: test/source/square.bc /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test/source/square.bc 
> cd /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test && /opt/gentoo/usr/lib/llvm/17/bin/clang-17 -c --offload-arch=gfx906 -emit-llvm -fgpu-rdc --gpu-bundle-output /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr/test/source/square.hip -o source/square.bc
> ```

It seems that the old hip version is causing the test build to fail. After I removed the old hip, there's no issue compiling comgr tests.

littlewu2508 commented:
6 tests failed for comgr:

79% tests passed, 6 tests failed out of 29                                                                                                                                                                                                         
                                                                                                                                                                                                                                                   
Total Test time (real) =   0.65 sec                                                                                                                                                                                                                
                                                                                                                                                                                                                                                   
The following tests FAILED:                                                                                                                                                                                                                        
         12 - comgr_compile_test (Subprocess aborted)                                                                                                                                                                                              
         13 - comgr_compile_minimal_test (Subprocess aborted)                                                                                                                                                                                      
         16 - comgr_compile_device_libs_test (Subprocess aborted)                                                                                                                                                                                  
         17 - comgr_compile_source_with_device_libs_to_bc_test (Subprocess aborted)                                                                                                                                                                
         28 - comgr_mangled_names_test (Subprocess aborted)                                                                                                                                                                                        
         29 - comgr_multithread_test (Subprocess aborted)

littlewu2508 commented:
`emerge hip` gives error at `src_config` on Gentoo Prefix system:

```
CMake Error at hipamd/CMakeLists.txt:484 (file):
  file COPY cannot find "/usr/bin/hipvars.pm": No such file or directory.
```

It seems that some absolute paths should be prefixified

littlewu2508 commented:
> Running `USE=test emerge =dev-libs/rocm-comgr-5.7.1` I got
> 
> ```
> FAILED: test/source/square.bc /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test/source/square.bc 
> cd /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr_build/test && /opt/gentoo/usr/lib/llvm/17/bin/clang-17 -c --offload-arch=gfx906 -emit-llvm -fgpu-rdc --gpu-bundle-output /run/user/18014/portage/dev-libs/rocm-comgr-5.7.1/work/ROCm-CompilerSupport-rocm-5.7.1/lib/comgr/test/source/square.hip -o source/square.bc
> ```

This does not only occurs when an old version of hip is installed. A up-to-date hip installation triggers it as well

littlewu2508 commented:
I got a strange file-not-found error when compiling rocBLAS:

```
[12/456] /opt/gentoo/usr/bin/x86_64-pc-linux-gnu-gfortran -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/roc
blas -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas/internal -O2 -pipe -O2 -pipe -Jclients/include/rocblas/internal -fpreprocessed -o clients/CMakeFiles/rocblas_fortran_client.dir/include/rocb
las_fortran_tensile.f90.o -c clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90-pp.f90                                                                                                                              
FAILED: clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90.o clients/include/rocblas/internal/rocblas_interface_tensile.mod                                                                                         
/opt/gentoo/usr/bin/x86_64-pc-linux-gnu-gfortran -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas -I/r
un/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas/internal -O2 -pipe -O2 -pipe -Jclients/include/rocblas/internal -fpreprocessed -o clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortr
an_tensile.f90.o -c clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90-pp.f90                                                                                                                                       
/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include/rocblas_fortran_tensile.f90:29:9:                                                                                                                        
                                                                                                                                                                                                                                                   
   29 |     use rocblas                                                                                                                                                                                                                            
      |         1                                                                                                                                                                                                                                  
Fatal Error: Cannot open module file ‘rocblas.mod’ for reading at (1): No such file or directory                                                                                                                                                   
compilation terminated.
```

littlewu2508 commented on sci-libs/rocBLAS/rocBLAS-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,109 @@
+IUSE="benchmark test"
+REQUIRED_USE="${ROCM_REQUIRED_USE}"
+
+RESTRICT="test" # Tests fail
```
I just observe only a trivial Fortran header include error mentioned at  https://github.com/gentoo/gentoo/pull/33400#issuecomment-1773730274. After copying headers from `clients/include/rocblas/internal` to `clients/include` at `${BUILD_DIR}`, everything if fine. I guess adding a include flag (probably an upstream bug) is the fix.

I passed all enabled tests on a single MI210 machine with `sci-libs/blis-0.9.0` as reference. OpenBLAS and netlib reference segfault somehow, probably they have bug.

littlewu2508 commented on sci-libs/rocRAND/rocRAND-5.7.1.ebuild:
```diff
@@ -0,0 +1,71 @@
+	addpredict /dev/dri/
+
+	# hipcc compiles everything with -O3; other modes cause clang backend errors
+	filter-flags -O?
```
I cannot reproduce `clang backend` errors. Maybe you should file a bug to llvm upstream. I would suggest put this as a ewarn, to remind people turning off `-O` flag manually if they encounter clang backend errors.

AngryLoki replied:
> I cannot reproduce `clang backend` errors.

Retried after system update (basically, clang 17.0.2 -> 17.0.3 and now it worked without removing -O2).
There is still an open question: is it a valid assumption, that it is good for end-user to use global CFLAGS for compiling GPU applications? It is a well-known fact, for example, that exactly -O3 mode provides the best result for CUDA in terms of registry pressure. Should not there be `GPU_CFLAGS_*` in pair to https://wiki.gentoo.org/wiki/CPU_FLAGS_* ?
Anyways, this is just an open question. For now the target is just to stabilize.

littlewu2508 replied:
>  Anyways, this is just an open question. For now the target is just to stabilize.

You are right, currently there's no detailed documentation on how to choose GPU optimization flag. Also, I did not find a way to set clang optimization level separately for CPU and GPU code generation. So currently let's strip the `-O` flag and use `-O3` which is the default set by hipcc

littlewu2508 replied:
> it outputs nothing

Currently clang does not properly detects the rocm installation located at `/usr` in Gentoo. So `hipcc` manually parsed `--rocm-path=${EPREFIX}/usr` in the perl script.

AngryLoki commented:
@littlewu2508 , I added https://github.com/gentoo/gentoo/pull/33400/commits/d2ed586f13de1a4628740e1d0619e48662cd0e5a - it should fix rocFFT test crash.

Do you have plans to apply https://github.com/llvm/llvm-project/pull/69842 as a patch for [compiler-rt-17.0.4](https://gitweb.gentoo.org/repo/gentoo.git/tree/sys-libs/compiler-rt/compiler-rt-17.0.4.ebuild)? This should unblock merge of this pull-request (as far as I understand it).

I was unable to reproduce file-not-found error when compiling rocBLAS, that you mentioned earlier. sci-libs/lapack-3.11 and x86_64-pc-linux-gnu-gfortran 13.2.1 20230826.

heroxbd commented:
Hi @AngryLoki , well done!  I have applied the hipcc patch as per https://github.com/gentoo/gentoo/pull/33845 and hip patches accumulated in this PR.

I have excluded the ISA compatibility patch though, because it makes more sense to Debian as a binary distribution to reduce package size.  But with Gentoo it is not a high priority.  Better leave it to upstream IMHO.

thesamesam commented:
@heroxbd 74964e47d7cb86d9cbb992f02e9de25a554727ac needed a revbump as it changed installed files/code (it's a runtime crash). But also, you've cherry-picked one(?) commit from this PR, and not included any reference to the PR. Please either merge the whole thing (are you merging things with `pram`?), or comment on this to ask the poster to rebase it and note what you cherry-picked.

heroxbd commented:
Sam James ***@***.***> writes:

> @heroxbd 74964e4 needed a revbump as it changed installed files/code
> (it's a runtime crash). 

Indeed, thanks.

> But also, you've cherry-picked one(?) commit from this PR, 

I picked several commits @AngryLoki authored on dev-util/hip and
squashed them into one.  Because some of them include reverts.

> and not included any reference to the PR. 

Is a line of
```
Bug: https://github.com/gentoo/gentoo/pull/33400
```
in the commit message enough?  Please enlighten me what is missing.

> Please either merge the whole thing (are you merging things with
> pram?),

Merged with plain git.

> or comment on this to ask the poster to rebase it and note what you
> cherry-picked.

Yesh, I realized it and added comments at
https://github.com/gentoo/gentoo/pull/33400#issuecomment-1817722581 .

thesamesam commented:
Yeah, I corrected my comment wrt reference.

heroxbd commented:
> I got a strange file-not-found error when compiling rocBLAS:
> 
> ```
> [12/456] /opt/gentoo/usr/bin/x86_64-pc-linux-gnu-gfortran -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/roc
> blas -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas/internal -O2 -pipe -O2 -pipe -Jclients/include/rocblas/internal -fpreprocessed -o clients/CMakeFiles/rocblas_fortran_client.dir/include/rocb
> las_fortran_tensile.f90.o -c clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90-pp.f90                                                                                                                              
> FAILED: clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90.o clients/include/rocblas/internal/rocblas_interface_tensile.mod                                                                                         
> /opt/gentoo/usr/bin/x86_64-pc-linux-gnu-gfortran -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include -I/run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas -I/r
> un/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1_build/include/rocblas/internal -O2 -pipe -O2 -pipe -Jclients/include/rocblas/internal -fpreprocessed -o clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortr
> an_tensile.f90.o -c clients/CMakeFiles/rocblas_fortran_client.dir/include/rocblas_fortran_tensile.f90-pp.f90                                                                                                                                       
> /run/user/18014/portage/sci-libs/rocBLAS-5.7.1-r1/work/rocBLAS-rocm-5.7.1/clients/include/rocblas_fortran_tensile.f90:29:9:                                                                                                                        
>                                                                                                                                                                                                                                                    
>    29 |     use rocblas                                                                                                                                                                                                                            
>       |         1                                                                                                                                                                                                                                  
> Fatal Error: Cannot open module file ‘rocblas.mod’ for reading at (1): No such file or directory                                                                                                                                                   
> compilation terminated.
> ```

Got the same problem.  But it works on the second run.  I guess it is a race condition in the build system.

AngryLoki commented:
@heroxbd , thanks for your help! I rebased remaining commits and squashed few of them.

Regarding "extend-isa" patches, I added it not for size reduction, but to extend supported set of hardware.

At this moment AMD provides options to build only for selected GPUs, described [here](https://rocm.docs.amd.com/en/latest/release/gpu_os_support.html#linux-supported-gpus). In general, it is possible to describe split targets in 3 groups

1. [Supported by AMD in ROCm](https://rocm.docs.amd.com/en/latest/release/gpu_os_support.html#linux-supported-gpus_
2. [Supported in the code of libraries](https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt) (Gentoo adds all of them in unofficial_amdgpu_targets)
3. [Supported as targets by LLVM/Clang](https://llvm.org/docs/AMDGPUUsage.html#processors) (Gentoo adds all of them in unofficial_amdgpu_targets) (Gentoo adds _some of them_ in unofficial_amdgpu_targets with patches)

|Architecture | Supported by AMD in ROCm | Supported by AMD (in rocBLAS code) | Also supported by LLVM/Clang|
| -- | --| -- | --|
|GCN3   |           | gfx803[1]              |  |
|GCN5.0 |           | gfx900[1]              | gfx902, gfx904  |
|GCN5.1 | gfx906    |                        | gfx909, gfx90c  |
|CDNA   | gfx908    |                        |   |
|CDNA2  | gfx90a    |                        |   |
|RDNA1  |           | gfx1010[1], gfx1012[1] | gfx1011[1], gfx1013  |
|RDNA2  | gfx1030   |                        | gfx1031[1], gfx1032, gfx1033, gfx1034, gfx1035, gfx1036  |
|RDNA3  | gfx1100[2]| gfx1101[1], gfx1102[1] | gfx1103  |

[1] Marked as unofficial_amdgpu_targets in [rocm.eclass](https://github.com/gentoo/gentoo/blob/master/eclass/rocm.eclass#L143C5-L143C12)
[2] Ubuntu 22.04 only binaries (and rocFFT will probably fail on gfx1100)

Additionally, there are `#if defined(__gfx803__)|| defined(__gfx900__) || ...` scattered accross the code ([example](https://github.com/ROCmSoftwarePlatform/rocFFT/blob/rocm-5.7.1/library/src/device/kernels/memory_gfx.h#L42C36-L46)), that prevent from passing arbitrary AMDGPU target. For some `unofficial_amdgpu_targets` patches were applied for such places. However, without having the proper hardware it is difficult to test, if all places were patched (for example, Gentoo has no patches for rocFFT, even though it uses such ifdefs). Additionally, some libraries (miopen) dynamically fetch target from `amdgpu-arch`-like calls, then try to find prebuilt kernel, then fail if not found.

There are group of targets, that are known to be compatible due to following [HSA specs](https://www.hsafoundation.com/). [I turned a $95 AMD APU into a 16GB VRAM GPU](https://old.reddit.com/r/Amd/comments/15t0lsm/i_turned_a_95_amd_apu_into_a_16gb_vram_gpu_and_it/) mentions that Ryzen 5 4600G CPU (gfx90c, GCN5.1) serves as a good replacement for 16GB GPU. Personally I use Ryzen 9 7950x3d (gfx1036, RDNA2) and can successfully run tests for ROCm packages. Therefore, as AMD user, I would like to offload computations to iGPU too.

There are 2 ways to achieve this in Gentoo:
1) Add patches that fallback to compatible kernels, when kernels for exact arch were not found (so that gfx1036 can specify gfx1030 as target and it will work).
2) Add all LLVM/Clang-supported targets to unofficial_amdgpu_targets.

This PR follows the first path.

The second way is probably more user-friendly for users, who specify a single target in AMDGPU_TARGETS. It is kind of hostile for those who specify the same targets of same architecture, because it basically only slows down compilation and produces larger fatbinaries, containing exactly the same code (because ISA is exactly the same). Additionally it requires more patches (for rocFFT and miOpen, for example). I can't estimate the amount of patches, but I expect it to grow with dependent software (pytorch? blender? onednn?). But it is not difficult: simple run with any iGPU identifies places that need to be patched.

What do you (and other, who read this) think?

If you still think that upstream can solve it, I could at least initiate discussion there (previous large discussion was in rocBLAS, but it should probably be in hip-runtime, I guess...)

heroxbd commented:
> Regarding "extend-isa" patches, I added it not for size reduction, but to extend supported set of hardware.

I am convinced by your explanation.  This nice little patch is interesting to Gentoo users, and bring us in line with Debian.  Thank you!

I have tested and merged many more packages including:  `dev-util/Tensile sci-libs/{hipBLAS,hipCUB,hipSPARSE,rocBLAS,rocPRIM,rocSOLVER,rocSPARSE,rocThrust}` .  The final round of merge will come after https://github.com/gentoo/gentoo/pull/33906 .  Have my respect and appreciation for all the work you have done.

littlewu2508 commented:
> but I need somebody with gfx1031 to confirm that migration is seamless

Probably a bad news, I am testing rocBLAS on gfx1031 now and it seems that there are tonnes of numerical errors. I'm not sure why this is happening, but I will try reverting the ISA hack, compile gfx1031 GPU code natively and re-run the tests to see what happens.

littlewu2508 commented:
> Probably a bad news, I am testing rocBLAS on gfx1031 now and it seems that there are tonnes of numerical errors. I'm not sure why this is happening

Sorry, I was on `::gentoo` master branch and it only contains the isa expansion patch `rocBLAS`, while patches for `hip` and `rocr-runtimr` are missing.

So I picked the `hip` and `rocr-runtime` commit from this PR and create https://github.com/gentoo/gentoo/pull/34260. After merging the isa expansion patch, my 6700XT passed all rocBLAS tests. Congratulations!

littlewu2508 commented on dev-libs/rccl/rccl-5.7.1.ebuild:
```diff
@@ -17,7 +17,7 @@ SLOT="0/$(ver_cut 1-2)"
 
 RDEPEND="
-	dev-util/hip
+	dev-util/hip:${SLOT}
```
Using sub slot to lock hip version here might be too strict, since the SLOT of hip-5.x.y is 5.x, meaning it restricts to a minor version. That is used to lock rocm-device-libs/rocm-comgr/hip, and rocBLAS/rocSOLVER, etc. From experiences it seems `sci-libs/*` and hip is compatible in the same major version.

Ref: https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=dde5dd2ff1162e22474fc87ccf6098943fee0caa

littlewu2508 replied:
I personally suggests using "=dev-util/hip-5*" here.

AngryLoki replied:
Ok, `=dev-util/hip-5*` seems to work fine.

Bratzmeister replied:
@littlewu2508 I had the same issue as you described here https://github.com/ROCm/ROCm-CompilerSupport/issues/53#issuecomment-1807354795 but in runtime not compile time with opencl. The linking error was also evident in output from clinfo command. unmerging and recompiling hip, rocm-comgr, rocm-device-libs, rocBLAS in order fixed the issue

littlewu2508 replied:
> hip, rocm-comgr, rocm-device-libs, rocBLAS

Thanks for the report, while the order confused me.

The issue is on opencl side, it should have nothing to do with hip and rocBLAS. So why the emerge order of hip and rocBLAS matters, and why hip should be in first place? hip should depend on rocm-device-libs and rocm-comgr.

Also, rocm-comgr should depend on rocm-device-libs. How did you manage to make a clean installation in rocm-comgr -> rocm-device-libs order?

Basically, rocm-comgr and rocm-device-libs is "part of llvm" but not upstreamed for some reasons, so they strongly correlated to llvm major versions, and their own version should also be consistent, and their order matters -- it's like you need first install/upgrade llvm and then clang.

Bratzmeister replied:
@littlewu2508 oh sorry I just realised now that my wording was misleading. the order was the unmerging order. However shortly after my report I noticed the same issue showing up again and I couldn't explain why. rocminfo again has the NULL errors.

I first installed rocm-device-libs and then rocm-comgr then lastly hip

negril commented:
Any chance you could include `https://github.com/ROCm/hipother` as well? So cuda can be used as backend. Would ease testing opencv/blender for me etc.

littlewu2508 commented:
> Any chance you could include https://github.com/ROCm/hipother as well? So cuda can be used as backend. Would ease testing opencv/blender for me etc.

Seems that this repo is only released in ROCm-6 and it still lacks documentation, so it won't be included in this PR.

See top notes on https://wiki.gentoo.org/wiki/ROCm. You can open a bug ticket and CC me on https://bugs.gentoo.org

Bratzmeister commented:
I´m really looking forward for full rocm 5.7.1 in gentoo so hopefully this PR will be merged soon. As with my RDNA3 I can't currently use rocm in gentoo (supported with 5.6 upwards) as stable is 5.3 and 5.7 is only partly available and I can't get any backend (opencl, hip, openmp) to work without errors or segfaults

AngryLoki commented:
@atupone 
I added all dependencies needed for https://bugs.gentoo.org/905286 , one small step for hip, one giant leap for dependencies :)

The only not updated component is now [dev-util/rocprofiler](https://packages.gentoo.org/packages/dev-util/rocprofiler). Unfortunately, old `no-aqlprofiler.patch` is completely defunct for 5.7.1, as rocprofiler is now tightly integrated with `libhsa-amd-aqlprofile64.so`. It is still technically possible to replace this lib with https://paste.gentoo.zip/rnLzgNYP , and it works, but there are no metrics other than "call started / call completed", which renders rocprofiler not too useful.

Bratzmeister commented:
with these changes it should be possible to incorporate rocm with pytorch right? https://bugs.gentoo.org/905325

littlewu2508 commented:
> with these changes it should be possible to incorporate rocm with pytorch right? https://bugs.gentoo.org/905325

Exatcly. I already have one https://github.com/littlewu2508/gentoo/commit/c12c96078cd02f7f38659460355dd022e8a07507 that works on my machine previously. I will update my build and re-test pytorch

AngryLoki commented:
Rebased after changes dev-util/cmake -> dev-build/cmake move.

> Exatcly. I already have one [littlewu2508@c12c960](https://github.com/littlewu2508/gentoo/commit/c12c96078cd02f7f38659460355dd022e8a07507) that works on my machine previously. I will update my build and re-test pytorch

I also have caffe2 2.1.2 here - https://github.com/gentoo/gentoo/commit/56aefda1a0bee3494c1d4fc844d53074fc404500. It is very similar, but I experience https://github.com/pytorch/pytorch/issues/103222 , so I patched it there.

> OpenBLAS and netlib reference segfault somehow, probably they have bug.

Btw, regarding this, could you check if you have nonstandard USE flags for blis or OpenBLAS (namely, 64bit-index)? I expect combination of `eselect-ldso` and `index-64bit` to crash due to https://github.com/OpenMathLib/OpenBLAS/issues/1763

mm1ke commented on dev-util/rocm-smi/rocm-smi-5.7.1.ebuild:
```diff
@@ -0,0 +1,62 @@
+
+DEPEND=""
+RDEPEND="${PYTHON_DEPS}"
+BDEPEND=""
```
empty variables could be removed (also `DEPEND`, `USE`)

mm1ke commented on dev-util/rocm-smi/rocm-smi-5.7.1.ebuild:
```diff
@@ -0,0 +1,62 @@
+	python_foreach_impl python_newscript python_smi_tools/rocm_smi.py rocm-smi
+	python_foreach_impl python_domodule python_smi_tools/rsmiBindings.py
+
+	mv "${ED}"/usr/share/doc/rocm_smi "${ED}"/usr/share/doc/${PN}-${PV}
```
missing `|| die`

thesamesam commented:
Can you give it a rebase for CI and I'll try take a look over the coming days? Thanks.

AngryLoki commented:
> Can you give it a rebase for CI and I'll try take a look over the coming days? Thanks.

Done, also added fix for https://bugs.gentoo.org/923986 (affects so that everything builds in clang/lld envoronment)

thesamesam commented on dev-cpp/frugally-deep/frugally-deep-0.15.30.ebuild:
```diff
@@ -0,0 +1,32 @@
+
+src_configure() {
+	local mycmakeargs=(
+		-DFDEEP_USE_DOUBLE=$(usex double)
```
I wonder if we really need USE=double here given nothing depends on it. If it's useful though, that's fine, it's just that historically, flags like this (double and such) have ended up breaking things when on/off, or wose, breaking ABI.

AngryLoki replied:
Removed `src_configure` completely. It makes no sense for header-only library (it only would make sense if somebody would include frugally-deep with `add_subdirectory`)

thesamesam commented on dev-cpp/functional-plus/functional-plus-0.2.22.ebuild:
```diff
@@ -0,0 +1,35 @@
+src_test() {
+	local BUILD_DIR="${WORKDIR}/${P}_build/test"
+	local CMAKE_USE_DIR="${S}/test"
+	cmake_src_configure
```
Not a fan of trying to do this in src_test. Can you conditionally do it in src_configure/src_compile instead?

AngryLoki replied:
Done! It looks much better now :)

thesamesam commented on dev-util/rocm-smi/rocm-smi-5.7.1.ebuild:
```diff
@@ -0,0 +1,59 @@
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_{9..12} )
```
9 is a noop nowadays, just make it 10..12.

thesamesam commented on dev-util/rocm-smi/rocm-smi-5.7.1.ebuild:
```diff
@@ -0,0 +1,59 @@
+)
+
+src_prepare() {
+	eapply_user
```
cmake_src_prepare already calls default, so you can drop eapply_user. User patches right now I think will be applied twice.

thesamesam commented on dev-util/rocm-smi/rocm-smi-5.7.1.ebuild:
```diff
@@ -0,0 +1,59 @@
+	python_foreach_impl python_newscript python_smi_tools/rocm_smi.py rocm-smi
+	python_foreach_impl python_domodule python_smi_tools/rsmiBindings.py
+
+	mv "${ED}"/usr/share/doc/rocm_smi "${ED}"/usr/share/doc/${PN}-${PV} || die
```
It should be PF, not PN-PV.

thesamesam commented on dev-util/roctracer/roctracer-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+	cmake_src_install
+
+	# remove unneeded copy
+	rm -r "${ED}/usr/share/doc/${PN}-${PV}-r1-asan" || die
```
This looks suspiciously precise. Why is it hardcoding the precise version, including revision?

AngryLoki replied:
Replaced with ${PF}

thesamesam commented on dev-util/roctracer/roctracer-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+	check_amdgpu
+	cd "${BUILD_DIR}" || die
+	# if LD_LIBRARY_PATH not set, dlopen cannot find correct lib
+	LD_LIBRARY_PATH="${EPREFIX}"/usr/lib64 bash run.sh || die
```
We shouldn't assume libdir is /usr/lib64, and we shouldn't overwrite it entirely:
```suggestion
	LD_LIBRARY_PATH="${EPREFIX}/usr/$(get_libdir):${LD_LIBRARY_PATH}" bash run.sh || die
```

thesamesam commented on sci-libs/composable-kernel/composable-kernel-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,81 @@
+
+src_test() {
+	check_amdgpu
+	export LD_LIBRARY_PATH="${BUILD_DIR}"/lib
```
Please prepend instead of overwriting LD_LIBRARY_PATH.

thesamesam commented on sci-libs/composable-kernel/composable-kernel-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,81 @@
+src_test() {
+	check_amdgpu
+	export LD_LIBRARY_PATH="${BUILD_DIR}"/lib
+	MAKEOPTS="-j1" cmake_src_test
```
cmake_src_test -j1 should work.

thesamesam commented on sci-libs/composable-kernel/files/composable-kernel-5.7.1-enable-test-examples.patch:
```diff
@@ -0,0 +1,19 @@
+--- a/CMakeLists.txt
```
Not a blocker for this specific PR but please send these patches upstream & include a link to the PR/commit at the top of the patch.

thesamesam commented on sci-libs/miopen/miopen-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,123 @@
+
+src_test() {
+	check_amdgpu
+	export LD_LIBRARY_PATH="${BUILD_DIR}"/lib
```
See other comments.

thesamesam commented on sci-libs/rocSPARSE/rocSPARSE-5.7.1-r1.ebuild:
```diff
@@ -45,7 +45,7 @@ IUSE="benchmark test"
 SLOT="0/$(ver_cut 1-2)"
 
-RDEPEND="dev-util/hip
+RDEPEND="=dev-util/hip-5*
```
This change needs a revbump (git mv is fine here).

thesamesam reviewed (changes requested):
Overall:
- Please also make sure the commit messages reflect why e.g. seds are being dropped and so on.
- It's not a blocker right now, but we really need to make an effort to try document and upstream these patches.

parona-source commented on sci-libs/composable-kernel/composable-kernel-5.7.1-r1.ebuild:
```diff
@@ -0,0 +1,81 @@
+		)
+	fi
+
+	CXX=hipcc cmake_src_configure
```
Its a good idea to strip unsupported flags for hipcc, for example a gcc user may not expect their gcc specific flags to be fed to clang here.

Setting CXX to hipcc and running `strip-unsupported-flags` should handle this.

```
ninja: job failed: /usr/bin/hipcc -D__HIP_PLATFORM_AMD__=1 -D__HIP_PLATFORM_HCC__=1 -I/var/tmp/portage/sci-libs/composable-kernel-5.7.1-r1/work/composable_kernel-rocm-5.7.1/profiler/include -I/var/tmp/portage/sci-libs/composable-kernel-5.7.1-r1/work/composable_kernel-rocm-5.7.1/library/include -I/var/tmp/portage/sci-libs/composable-kernel-5.7.1-r1/work/composable_kernel-rocm-5.7.1/include  -O3 -march=znver2 -pipe -flto=auto -fuse-linker-plugin -frecord-gcc-switches -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing -DNDEBUG -std=c++17   -Wall -Wextra -Wcomment -Wendif-labels -Wformat -Winit-self -Wreturn-type -Wsequence-point -Wswitch -Wtrigraphs -Wundef -Wuninitialized -Wunreachable-code -Wunused -Wno-reserved-identifier -Werror -Wsign-compare -Wno-extra-semi-stmt -Wno-missing-field-initializers -Wno-deprecated-declarations -Wall -Wextra -Wcomment -Wendif-labels -Wformat -Winit-self -Wreturn-type -Wsequence-point -Wswitch -Wtrigraphs -Wundef -Wuninitialized -Wunreachable-code -Wunused -Wno-reserved-identifier -Werror -Wsign-compare -Wno-extra-semi-stmt -Weverything -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-conversion -Wno-double-promotion -Wno-exit-time-destructors -Wno-extra-semi -Wno-float-conversion -Wno-gnu-anonymous-struct -Wno-gnu-zero-variadic-macro-arguments -Wno-missing-prototypes -Wno-nested-anon-types -Wno-padded -Wno-return-std-move-in-c++11 -Wno-shorten-64-to-32 -Wno-sign-conversion -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-weak-vtables -Wno-covered-switch-default -Wno-unsafe-buffer-usage -Werror -Weverything -Wno-global-constructors -x hip --offload-arch=gfx1010 -MD -MT profiler/src/CMakeFiles/ckProfiler.dir/profile_conv_fwd_bias_relu_add.cpp.o -MF profiler/src/CMakeFiles/ckProfiler.dir/profile_conv_fwd_bias_relu_add.cpp.o.d -o profiler/src/CMakeFiles/ckProfiler.dir/profile_conv_fwd_bias_relu_add.cpp.o -c /var/tmp/portage/sci-libs/composable-kernel-5.7.1-r1/work/composable_kernel-rocm-5.7.1/profiler/src/profile_conv_fwd_bias_relu_add.cpp
clang++: error: optimization flag '-fuse-linker-plugin' is not supported [-Werror,-Wignored-optimization-argument]
clang++: error: optimization flag '-fuse-linker-plugin' is not supported [-Werror,-Wignored-optimization-argument]
```

parona-source replied:
> https://github.com/gentoo/gentoo/pull/34869#issuecomment-1954823915

So to clarify a best way to get around this is to have `CXX="clang" strip-unsupported-flags` in the ebuild. it will strip all the clang unsupported flags  that hipcc would feed to it later.

Bratzmeister commented:
are there plans to include https://github.com/ROCm/rocWMMA/releases in the gentoo rocm stack? it has supported releases for 5.7.1 as well

a-repko commented on dev-util/Tensile/Tensile-5.7.1-r1.ebuild:
```diff
@@ -27,7 +27,7 @@ RDEPEND="${PYTHON_DEPS}
 	>=dev-cpp/msgpack-cxx-6.0.0
```
missing dependency
```suggestion
	client? ( dev-libs/boost )
	>=dev-cpp/msgpack-cxx-6.0.0
```

AngryLoki replied:
Fixed, thanks!

a-repko commented on sci-libs/rocFFT/rocFFT-5.7.1-r1.ebuild:
```diff
@@ -29,7 +29,7 @@ perfscripts? (
 ${PYTHON_DEPS}"
 
-DEPEND="dev-util/hip
+DEPEND="=dev-util/hip-5*
```
missing dependency (boost was missing already from the version 5.1.3)
```suggestion
DEPEND="=dev-util/hip-5*
	benchmark? (
		dev-libs/boost
		sci-libs/hipRAND
	)
```

a-repko replied:
Also, I have the following additional proposal for your `rocBLAS-5.7.1` ebuild, which I didn't know how to include directly. All these fixes are intended to resolve build errors with specific USE flags.
```patch
--- sci-libs/rocBLAS/rocBLAS-5.7.1-r1.ebuild.AngryLoki
+++ sci-libs/rocBLAS/rocBLAS-5.7.1-r2.ebuild
@@ -4,7 +4,7 @@
 EAPI=8
 
 DOCS_BUILDER="doxygen"
-DOCS_DIR="docs"
+DOCS_DIR="docs/.doxygen"
 DOCS_DEPEND="media-gfx/graphviz"
 ROCM_VERSION=${PV}
 inherit cmake docs edo multiprocessing rocm
@@ -57,6 +57,11 @@
 }
 
 src_configure() {
+	if [[ ! -d /usr/include/boost ]]; then
+		# the case of dev-cpp/msgpack-cxx[-boost] with dev-util/Tensile
+		append-cxxflags -DMSGPACK_NO_BOOST
+	fi
+
 	addpredict /dev/random
 	addpredict /dev/kfd
 	addpredict /dev/dri/
```
BTW, thank you for the new and updated packages. They work well for me, as far as I could check (also after including `boost`, of course).

negril replied:
> which I didn't know how to include directly.

You can file a PR against https://github.com/AngryLoki/gentoo/tree/rocm-5.7 . The path needs a `${EPREFIX}` prepended and you need to explain the `DOCS_DIR` change.

Bratzmeister replied:
> BTW, thank you for the new and updated packages. They work well for me, as far as I could check (also after including `boost`, of course).

same here. Finally ROCm is in a workable state for me (gfx1100 user)

a-repko replied:
I tested all the merged ebuilds, and they work OK. The OpenCL issue is still present (of course, whole ROCm was rebuilt with the same toolchain), but the segfault is buried deep in /usr/lib64/libamdocl64.so (perhaps it is a problem of the client application called gpuowl, I didn't tested anything else; and clinfo doesn't have any problems) and USE=debug won't give more details in gdb, so let's ignore it for now (rocm-opencl-runtime-5.5.1 is working well for me).

In the meantime, I (hopefully) found a fix for an older bug in rocSPARSE, see https://bugs.gentoo.org/924446

Also, please add missing dependency in rocWMMA:
rocWMMA[test] depends on rocBLAS (during configure step, and probably also later)

I'm quite happy that fixed ROCm 5.7.1 works also on APU (Ryzen Pro 5750G). There are some freezes (maybe due to currently testing it on a text console only), but they can be easily "unfrozen" by simultaneously running rocPRIM_benchmark_warp_exchange. Of course, a dedicated GPU doesn't have such problems.

Bratzmeister commented:
> are there plans to include https://github.com/ROCm/rocWMMA/releases in the gentoo rocm stack? it has supported releases for 5.7.1 as well

thanks https://github.com/gentoo/gentoo/pull/33400/commits/9d8875fb4543c27f4e4f48c08391c788925d17e3
